Add TFM support for stm32l562e_dk board

yestin opened the pull request:
This PR is to add the TF-M support for the stm32l562e_dk board.

yestin committed:
boards: stm32l562e_dk: Add non secure target

This commit adds a second target for the stm32l562e_dk board.
The non secure target can be configured for TFM IPC application.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin committed:
boards: stm32l562e_dk: Enable TF-M IPC application

This commit enables the TF-M IPC sample application on stm32l562e_dk
board.
It provides device tree flash partition as an overlay in order to
configure and flash the bl2, secure/non secure firmwares.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin committed:
boards: stm32l562e_dk: Avoid GPIO pinctrl for non secure target

After TZ is enabled, the GPIO peripherals are secured and are
not accessible from non secure world.
This commit prevents the GPIO pinctrl from the non secure target
for stm32l562e_dk board.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin commented on boards/arm/stm32l562e_dk/stm32l562e_dk_ns.dts:
```diff
@@ -15,8 +15,8 @@
-		zephyr,console = &usart1;
-		zephyr,shell-uart = &usart1;
+		zephyr,console = &lpuart1;
+		zephyr,shell-uart = &lpuart1;
```
This PR works but is not ideal. After TZ is enabled, the log output over usart1 is garbage from the non secure Zephyr application. I have used lpuart1 as a workaround but it is a bit awkward because I have to jump wire to wire out the lpuart1 TX/RX pins on stm32l562e_dk board, and the logs for mcuboot and secure firmware come out of usart1 while the log for non secure application comes out of lpuart1.

I have tried to avoid the GPIOA pin control in non secure target since GPIOA is a Trustzone protected peripheral after TZEN is enabled.

I ran the TFM sample application from the STM32Cube_FW_L5 package, the logging seems to work in non secure image, but I don't see anything special in the UART(COM) port initialization code.

After I switch to lpuart1, the logging in non secure application is working, but I don't understand the different since both lpuart1 and usart1 are securable peripherals, and are non-secure after reset.

I have spent quite some time to debug this issue but no avail so far. Maybe can @erwango @ABOSTM shed some light on this?

erwango replied:
@yestin from what I see this is specific to this dk board and this works on the nucleo. I guess this is due to console confugration for this board on TFM side. Maybe @jamike you can confirm this ?

erwango replied:
@yestin just to be sure: have you tried to keep uart1 as NS console **without pinmuxing enabled**

yestin replied:
I believe this commit ([84d97dd8](https://github.com/zephyrproject-rtos/zephyr/pull/31118/commits/84d97dd894571bb0993c437e8effd71c69242f38)) shall disable the pinmuxing for the uart1 as NS console. After build, I checked the generated dts file build/zephyr/zephyr.dts and did not see the pinmux configuration.

erwango replied:
> I believe this commit ([84d97dd8](https://github.com/zephyrproject-rtos/zephyr/pull/31118/commits/84d97dd894571bb0993c437e8effd71c69242f38)) shall disable the pinmuxing for the uart1 as NS console. After build, I checked the generated dts file build/zephyr/zephyr.dts and did not see the pinmux configuration.

And does it allow to get working console on uart1 ?

erwango replied:
ok, I had a test and I confirm I'm seeing the same thing.

erwango replied:
Though, at first sight, I'd rather suspect a clock issue.

erwango replied:
So I manage to make ns console work on uart1 by setting `current-speed = <57600>;`.
This confirms it is a clock issue:
TFM and Zephyr doesn't configure the same clock tree for usart1.
Since we don't face this issue on lpuart1, I suspect the misalignment either comes from APB2 PRESC or USART1SEL(RCC_BDCR).

yestin replied:
This is good stuff. I was able to see the ns console working with your suggestion.

Unfortunately, I think I lost my stm32l562e_dk board when I was trying to deactivate TZEN. I was able to recover it last time by applying 3V3 on PH3-BOOT0 signal to make it boot from RSS, but not this time. 😞 

I guess setting RDP level to 1 before TZEN deactivation was not a good idea. Anyway, I will not be able to debug any further on this for some time.

erwango replied:
I faced some issue with the board myself during the investigation.
I managed to get it back using stm32cubeprogrammer, by erasing the flash fully and then programming a hello world (using stm32cubeprogrammer UI).

erwango replied:
Should be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/31178

yestin replied:
Thanks for the fix. I have rebased on top of your fix and verified the new patch that the logging works well on usart1 now.

erwango commented on samples/tfm_integration/tfm_ipc/boards/stm32l562e_dk_ns.overlay:
```diff
@@ -0,0 +1,55 @@
+  * - TFM_PSA_API=ON (IPC)
+  * - ISOLATION_LEVEL 2
+  * - TEST_S=ON (REGRESSION)
+  * - TEST_NS=ON (REGRESSION)
```
After some search it seems that this is not the case anymore:
https://github.com/zephyrproject-rtos/zephyr/pull/30210

It might be worth checking potential impacts on partition sizes

yestin replied:
I see. And yes, the non secure image in this sample is pretty small, like 19KB. I can try to decrease the non secure image size, but may still not enough for the secondary partitions.

microbuilder replied:
I was just looking at this PR, and indeed with 512 KB flash it will be a bit of a juggling-act to try to make this work with the dual-bank bootloader. It's possible, but more complex projects will be a challenge without using a different approach for the secure bootloader.

yestin replied:
In the new patch, I leave the partition layout as is except updated the comments. As the secure image is almost 224 KB, it could not fit a secondary partition in 512 KB flash.

ioannisg reviewed (approved):
porting looks fine to me, and along the lines of the other tf-m capable platforms.
needs formal approval from @erwango

galak commented:
@yestin do you mind rebasing this to address the merge conflict.

ioannisg commented:
@yestin may i kindly ask you to do one more rebase of this PR? I recently merged #32044 which centralizes the signing for TFM builds, so you do not need to duplicate this any more in the board definition. Please, take a look at the other TFM-capable boards, for guidance.

After this rebase, assuming I get a +1 from @microbuilder , I'll merge this PR.

yestin commented:
> @yestin may i kindly ask you to do one more rebase of this PR? I recently merged #32044 which centralizes the signing for TFM builds, so you do not need to duplicate this any more in the board definition. Please, take a look at the other TFM-capable boards, for guidance.
> 
> After this rebase, assuming I get a +1 from @microbuilder , I'll merge this PR.

Yes, I will look into rebasing on top of the new changes. I am running into issue that the signing generates .hex file instead .bin file with the new merge. The TFM_UPDATE.sh script calls STM32_Programmer_CLI to flash, and when the download file is hex file, it does not honor the downloading address that passed in.

The .hex file for NS image still links against the start address at 0x08000000, which is not a correct start address in secure state. For stm32l552xx and stm32l562xx, the flash memory space is aliased to 0x0C000000 in secure state (TZEN=1). I still need some time to debug this.

yestin committed:
modules: trusted-firmware-m: Allow to adjust hex file base address

This commit allows to append an optional --hex-addr argument to
the wrapper script if speficied. This can adjust the base address
of the output hex file when signing the non-secure or secure
firmware images.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin committed:
boards: nucleo_l552ze_q: Leverage consolidated TFM signing code

This commit removes the signing commands from the post build steps,
in order to leverage the consolidated TFM signing code.

Also with the support to adjust the hex base address when signing,
there is no need to run the TFM_UPDATE.sh script. We can use west
flash to flash the merged hex file on the board.

Signed-off-by: Yestin Sun <sunyi0804@gmail.com>

yestin commented:
@ioannisg @microbuilder the rebased changes are ready for review. I add an option to the signing script to adjust the hex base address. This can allow the output hex file to be flashed to a different address. In the case for nucleo_l552ze_q and stm32l562e_dk boards, they have a different flash memory space (0x0C000000) when TZ is enabled.

I also update the nucleo_l552ze_q board to leverage the centralized signing code. cc: @erwango

yestin commented:
@ioannisg do you get a chance to review the PR? It has been pending for some time and I am worried I would have to rebase it again. Please let me know if you have concerns.

ioannisg commented:
> @ioannisg do you get a chance to review the PR? It has been pending for some time and I am worried I would have to rebase it again. Please let me know if you have concerns.

@yestin apologies for forgetting this PR, for some time. To clarify, the reason for adding DNM flag is that i was not able to run the tfm_ipc test properly, on the STM32L562 board, using the PR, so needed to investigate this more.

@yestin are you able to successfully compile, flash and run the tfm_ipc sample on this board _and_ on nucleo_l552ze_q?

erwango commented on boards/arm/nucleo_l552ze_q/doc/nucleol552ze_q.rst:
```diff
@@ -276,7 +276,16 @@ You should see the following message on the console:
+using the ``nucleo_l552ze_q_ns`` target. When building a ``*_ns`` image with TF-M,
+a ``build/tfm/install/postbuild.sh`` bash script will be run as a post-build step
+to make some required flash layout changes. The ``build/tfm/regression.sh`` script
+will need to be run to perform device initialization, and then run ``west flash --hex-file build/tfm_merged.hex``
```
Can you review the command here?
I had to go in `samples/tfm_integration/tfm_ipc/build` and then launch `west flash --hex-file tfm_merged.hex`

ioannisg replied:
Yes, this needs an improvement.

yestin replied:
Looks like the build output is generated to the sample project directory instead of the root directory. I did a clean run and can see the merged hex file is generated at `build/tfm_merged.hex`. Do you have any local environments that may impact?

erwango replied:
@yestin the file is present here indeed, though, when running the command, I'm getting the following west error:
```
$ west flash --hex-file build/stm32l562e_dk_ns/tfm_merged.hex
FATAL ERROR: --build-dir was not given, /local/mcu/zephyrproject/zephyr/samples/tfm_integration/tfm_ipc is not a build directory and the default build directory cannot be determined. Check your build.dir-fmt configuration option
```
@ioannisg is that what you're seeing as well ?

erwango replied:
@yestin I confirm this is only due to my local build.dir-fmt configuration.
I've raised https://github.com/zephyrproject-rtos/zephyr/issues/33498 to discuss it

erwango reviewed (approved):
Approved, I'll review the doc part once merged.

ioannisg commented on boards/arm/stm32l562e_dk/stm32l562e_dk_ns_defconfig:
```diff
@@ -0,0 +1,44 @@
+CONFIG_UART_CONSOLE=y
+
+# Enable MPU
+CONFIG_ARM_MPU=y
```
We should probably enable HW_STACK_PROTECTION=y here as well. Or is there a problem with this @yestin @erwango ?

yestin commented:
> > @ioannisg do you get a chance to review the PR? It has been pending for some time and I am worried I would have to rebase it again. Please let me know if you have concerns.
> 
> @yestin apologies for forgetting this PR, for some time. To clarify, the reason for adding DNM flag is that i was not able to run the tfm_ipc test properly, on the STM32L562 board, using the PR, so needed to investigate this more.
> 
> @yestin are you able to successfully compile, flash and run the tfm_ipc sample on this board _and_ on nucleo_l552ze_q?

Yes, I verify that I can run the tfm_ipc sample on both boards.
